### PR TITLE
feat: update taint nodes by condition to GA

### LINKED ISSUE #82635: Graduate TaintNodesByCondition to GA
This feature has been in Beta since 1.12, we can now graduate it to GA.

/sig scheduling
/cc @k82cn

### PR DESCRIPTION
**Pull request by draveness**
/kind feature
/sig scheduling
/priority important-soon
/assign @k82cn 

**What this PR does / why we need it**:

Graduate TaintNodesByCondition to GA

**Which issue(s) this PR fixes**:

Fixes #82635

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Graduate TaintNodesByCondition to GA in 1.17. (feature gate will be removed in 1.18) action required
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -135,6 +135,7 @@ const (
 
 	// owner: @k82cn
 	// beta: v1.12
+	// GA: v1.17
```
Comment (k82cn) [inline pkg/features/kube_features.go]: from GA feature, we just remove it :)

Comment (k82cn) [reply pkg/features/kube_features.go]: and for this feature, it's also to remove some predicates about node condition.

**Comment by draveness** [comment]
The node e2e integration tests are failing since we are removing the TaintNodeByCondition feature gates from the codebase.

https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82703/pull-kubernetes-node-e2e/1173534864813592576/

I scanned through the PR history and found out TaintNodeByCondition was disabled on purpose in node e2e test. https://github.com/kubernetes/kubernetes/pull/73097#issuecomment-457390730

cc/ @bsalamat @liggitt @yujuhong, do you have any idea on how to fix this?

**Comment by liggitt** [comment]
If the node e2e doesn't want to run the node controller (which handles resolving this taint if I remember correctly), the node e2e will need to mimic that behavior in its setup.

In general, running partial environments in e2e tests is not recommended, because of the duplication/drift inherent in running a system different than a "real" cluster.

File: pkg/features/kube_features.go (hunk 1 above)
Comment (draveness) [reply pkg/features/kube_features.go]: Please take another look, thanks!

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -526,7 +519,6 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	EphemeralContainers:                 {Default: false, PreRelease: featuregate.Alpha},
 	PodShareProcessNamespace:            {Default: true, PreRelease: featuregate.Beta},
 	PodPriority:                         {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.18
-	TaintNodesByCondition:               {Default: true, PreRelease: featuregate.Beta},
```
Comment (wojtek-t) [inline pkg/features/kube_features.go]: If we're going to move this feature to GA, we shouldn't be removing the feature gate in the same release.

Comment (draveness) [reply pkg/features/kube_features.go]: I removed the feature gates because of Klaus's https://github.com/kubernetes/kubernetes/pull/82703#discussion_r324420293 here...

cc/ @k82cn

File: pkg/features/kube_features.go (hunk 1 above)
Comment (wojtek-t) [reply pkg/features/kube_features.go]: No - we don't remove feature gates for GA (at least not immediately - otherwise if someone is setting up feature gate, we will break them).

Comment (liggitt) [reply pkg/features/kube_features.go]: we *do* lock to true when promoting to GA (`LockToDefault: true` in the feature gate setup), and remove code that checks the gate

Comment (draveness) [reply pkg/features/kube_features.go]: Addressed

**Comment by draveness** [comment]
/retest

Regards,
Draven

File: pkg/features/kube_features.go (hunk 1 above)
Comment (k82cn) [reply pkg/features/kube_features.go]: @liggitt , @wojtek-t , when should remove the function code feature? Only keep feature gate for one release :)

And good to know we did not remove it immediately :)

Comment (liggitt) [reply pkg/features/kube_features.go]: one release. announce graduation to GA and feature gate deprecation in 1.17, then remove in 1.18

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 3]
```diff
@@ -62,34 +62,30 @@ func defaultPredicates() sets.String {
...
-		factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.CheckNodeUnschedulablePred)
-
-		klog.Infof("TaintNodesByCondition is enabled, PodToleratesNodeTaints predicate is mandatory")
-	}
+	// Remove "CheckNodeCondition", "CheckNodeMemoryPressure", "CheckNodePIDPressure"
+	// and "CheckNodeDiskPressure" predicates
+	factory.RemoveFitPredicate(predicates.CheckNodeConditionPred)
+	factory.RemoveFitPredicate(predicates.CheckNodeMemoryPressurePred)
+	factory.RemoveFitPredicate(predicates.CheckNodeDiskPressurePred)
+	factory.RemoveFitPredicate(predicates.CheckNodePIDPressurePred)
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: It is not under a feature gate anymore, for those, we should just remove there registration from 
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/defaults/register_predicates.go#L106-L116

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 4]
```diff
@@ -62,34 +62,30 @@ func defaultPredicates() sets.String {
...
+	// Remove key "CheckNodeCondition", "CheckNodeMemoryPressure", "CheckNodePIDPressure" and "CheckNodeDiskPressure"
+	// from ALL algorithm provider
+	// The key will be removed from all providers which in algorithmProviderMap[]
+	// if you just want remove specific provider, call func RemovePredicateKeyFromAlgoProvider()
+	factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodeConditionPred)
+	factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodeMemoryPressurePred)
+	factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodeDiskPressurePred)
+	factory.RemovePredicateKeyFromAlgorithmProviderMap(predicates.CheckNodePIDPressurePred)
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: instead of doing this, remove them from the default predicates list: https://github.com/kubernetes/kubernetes/blob/b8d8f1a35e4d3b1eb37557c91e3522ee68618585/pkg/scheduler/algorithmprovider/defaults/defaults.go#L41

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 5]
```diff
@@ -62,34 +62,30 @@ func defaultPredicates() sets.String {
...
+
+	// Fit is determined based on whether a pod can tolerate all of the node's taints
+	factory.RegisterMandatoryFitPredicate(predicates.PodToleratesNodeTaintsPred, predicates.PodToleratesNodeTaints)
+	// Fit is determined based on whether a pod can tolerate unschedulable of node
+	factory.RegisterMandatoryFitPredicate(predicates.CheckNodeUnschedulablePred, predicates.CheckNodeUnschedulablePredicate)
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: add them to predicate registration: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/defaults/register_predicates.go

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 6]
```diff
@@ -62,34 +62,30 @@ func defaultPredicates() sets.String {
...
+	// Insert Key "PodToleratesNodeTaints" and "CheckNodeUnschedulable" To All Algorithm Provider
+	// The key will insert to all providers which in algorithmProviderMap[]
+	// if you just want insert to specific provider, call func InsertPredicateKeyToAlgoProvider()
+	factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.PodToleratesNodeTaintsPred)
+	factory.InsertPredicateKeyToAlgorithmProviderMap(predicates.CheckNodeUnschedulablePred)
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: add them to the default predicates list.

File: pkg/features/kube_features.go (hunk 1 above)
Comment (k82cn) [reply pkg/features/kube_features.go]: got that, thanks very much :)

File: pkg/scheduler/algorithmprovider/defaults/register_predicates.go [hunk 7]
```diff
@@ -103,20 +103,11 @@ func init() {
 	// (e.g. kubelet and all schedulers)
 	factory.RegisterFitPredicate(predicates.GeneralPred, predicates.GeneralPredicates)
 
-	// Fit is determined by node memory pressure condition.
-	factory.RegisterFitPredicate(predicates.CheckNodeMemoryPressurePred, predicates.CheckNodeMemoryPressurePredicate)
```
Comment (draveness) [inline pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: If we remove this, it would break backward compatibility. https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82703/pull-kubernetes-bazel-test/1174606131515887619

/cc @ahg-g

Comment (draveness) [reply pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: The latest commit https://github.com/kubernetes/kubernetes/pull/82703/commits/f642bc2feb755cb6f834787163725a498cda4dce fix the failing test

Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: This predicate was effectively removed throughout beta, so it seems the compatibility test didn't take into account feature gates? in any case, I think we should keep them registered to maintain backward compatibility, add a comment to mark them as deprecated and just remove them from the default list.

Comment (draveness) [reply pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: But we have to register `CheckNodeConditionPred` as a mandatory fit predicate and remove it when apply the feature gates

Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: does CheckNodeConditionPred have to be registered as mandatory to keep compatibility? I think doing ```RegisterFitPredicate(predicates.CheckNodeConditionPred, predicates.CheckNodeConditionPredicate)``` would be enough.

Comment (draveness) [reply pkg/scheduler/algorithmprovider/defaults/register_predicates.go]: > does CheckNodeConditionPred have to be registered as mandatory to keep compatibility? I think doing `RegisterFitPredicate(predicates.CheckNodeConditionPred, predicates.CheckNodeConditionPredicate)` would be enough.

What about the latter one?

```go
--- FAIL: TestCompatibility_v1_Scheduler (0.01s)
    compatibility_test.go:1216: 1.16: Expected:
        	api.Policy{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, Predicates:[]api.PredicatePolicy{api.PredicatePolicy{Name:"MatchNodeSelector", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodFitsResources", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodFitsHostPorts", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"HostName", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"NoDiskConflict", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"NoVolumeZoneConflict", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodToleratesNodeTaints", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxEBSVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxGCEPDVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxAzureDiskVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxCSIVolumeCountPred", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxCinderVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MatchInterPodAffinity", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"GeneralPredicates", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckVolumeBinding", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckNodeUnschedulable", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"TestServiceAffinity", Argument:(*api.PredicateArgument)(0xc0002ca7c0)}, api.PredicatePolicy{Name:"TestLabelsPresence", Argument:(*api.PredicateArgument)(0xc0002ca7e0)}}, Priorities:[]api.PriorityPolicy{api.PriorityPolicy{Name:"EqualPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"ImageLocalityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"LeastRequestedPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"BalancedResourceAllocation", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"SelectorSpreadPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"NodePreferAvoidPodsPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"NodeAffinityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"TaintTolerationPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"InterPodAffinityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"MostRequestedPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"RequestedToCapacityRatioPriority", Weight:2, Argument:(*api.PriorityArgument)(0xc000179840)}}, ExtenderConfigs:[]api.ExtenderConfig{api.ExtenderConfig{URLPrefix:"/prefix", FilterVerb:"filter", PreemptVerb:"", PrioritizeVerb:"prioritize", Weight:1, BindVerb:"bind", EnableHTTPS:true, TLSConfig:(*api.ExtenderTLSConfig)(0xc000154b40), HTTPTimeout:1, NodeCacheCapable:true, ManagedResources:[]api.ExtenderManagedResource{api.ExtenderManagedResource{Name:"example.com/foo", IgnoredByScheduler:true}}, Ignorable:true}}, HardPodAffinitySymmetricWeight:0, AlwaysCheckAllPredicates:false}
        Got:
        	api.Policy{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, Predicates:[]api.PredicatePolicy{api.PredicatePolicy{Name:"MatchNodeSelector", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodFitsResources", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodFitsHostPorts", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"HostName", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"NoDiskConflict", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"NoVolumeZoneConflict", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"PodToleratesNodeTaints", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckNodeMemoryPressure", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckNodeDiskPressure", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckNodePIDPressure", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckNodeCondition", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxEBSVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxGCEPDVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxAzureDiskVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxCSIVolumeCountPred", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MaxCinderVolumeCount", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"MatchInterPodAffinity", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"GeneralPredicates", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"CheckVolumeBinding", Argument:(*api.PredicateArgument)(nil)}, api.PredicatePolicy{Name:"TestServiceAffinity", Argument:(*api.PredicateArgument)(0xc00045fc30)}, api.PredicatePolicy{Name:"TestLabelsPresence", Argument:(*api.PredicateArgument)(0xc00045fc60)}}, Priorities:[]api.PriorityPolicy{api.PriorityPolicy{Name:"EqualPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"ImageLocalityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"LeastRequestedPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"BalancedResourceAllocation", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"SelectorSpreadPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"NodePreferAvoidPodsPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"NodeAffinityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"TaintTolerationPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"InterPodAffinityPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"MostRequestedPriority", Weight:2, Argument:(*api.PriorityArgument)(nil)}, api.PriorityPolicy{Name:"RequestedToCapacityRatioPriority", Weight:2, Argument:(*api.PriorityArgument)(0xc000518ca0)}}, ExtenderConfigs:[]api.ExtenderConfig{api.ExtenderConfig{URLPrefix:"/prefix", FilterVerb:"filter", PreemptVerb:"", PrioritizeVerb:"prioritize", Weight:1, BindVerb:"bind", EnableHTTPS:true, TLSConfig:(*api.ExtenderTLSConfig)(0xc0005cd710), HTTPTimeout:1, NodeCacheCapable:true, ManagedResources:[]api.ExtenderManagedResource{api.ExtenderManagedResource{Name:"example.com/foo", IgnoredByScheduler:true}}, Ignorable:true}}, HardPodAffinitySymmetricWeight:0, AlwaysCheckAllPredicates:false}
    compatibility_test.go:1252: Registered predicates are missing from compatibility test (add to test stanza for version currently in development): []string{"CheckNodeUnschedulable"}
```

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 8]
```diff
@@ -1194,6 +1194,133 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 				}},
 			},
 		},
+		"1.17": {
```
Comment (ahg-g) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: Just heads up that you will need to update this after https://github.com/kubernetes/kubernetes/pull/82735 is merged.

**Review by ahg-g — COMMENTED** [review]
/lgtm 

for scheduler changes

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 8 above)
Comment (draveness) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: done

**Comment by draveness** [comment]
Hi @wojek-t, all the comments have been addressed, could you please take a look at this PR?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 9]
```diff
@@ -467,9 +466,7 @@ func NewNodeLifecycleController(
 		}),
 	})
 
-	if nc.taintNodeByCondition {
-		klog.Infof("Controller will taint node by condition.")
-	}
+	klog.Infof("Controller will taint node by condition.")
```
Comment (k82cn) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: do we still need this log?

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 10]
```diff
@@ -47,12 +47,9 @@ func defaultPredicates() sets.String {
 		predicates.MatchInterPodAffinityPred,
 		predicates.NoDiskConflictPred,
 		predicates.GeneralPred,
-		predicates.CheckNodeMemoryPressurePred,
-		predicates.CheckNodeDiskPressurePred,
-		predicates.CheckNodePIDPressurePred,
-		predicates.CheckNodeConditionPred,
 		predicates.PodToleratesNodeTaintsPred,
```
Comment (k82cn) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: I think `PodToleratesNodeTaintsPred` should be mandatory; if user enable TaintNodeByCondition without this predicates in configuration, will we change the behavior?

BTW, should we also identify mandatory predicates to make sure scheduler always "work".

Comment (k82cn) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: > I think PodToleratesNodeTaintsPred should be mandatory; if user enable TaintNodeByCondition without this predicates in configuration, will we change the behavior?

Ops, never mind; just see other diffs handled that.

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 11]
```diff
@@ -294,11 +294,11 @@ type Controller struct {
...
 	useTaintBasedEvictions bool
 
-	// if set to true, NodeController will taint Nodes based on its condition for 'NetworkUnavailable',
-	// 'MemoryPressure', 'PIDPressure' and 'DiskPressure'.
-	taintNodeByCondition bool
-
 	nodeUpdateQueue workqueue.Interface
+
+	// if set to true, Controller will skip taint nodes by condition.
+	// NOTE: used in scheduler integration test.
```
Comment (k82cn) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: please also add a comment on when should we remove that field :)

**Comment by k82cn** [comment]
LGTM overall, just minor comments :)

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 9 above)
Comment (draveness) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: removed

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 11 above)
Comment (draveness) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: done

**Comment by draveness** [comment]
> LGTM overall, just minor comments :)

Thanks for the review, all the comments have been addressed.

**Comment by draveness** [comment]
Hi @wojtek-t @k82cn, could you please take another look? Thanks

**Comment by ahg-g** [comment]
@wojtek-t @deads2k can you take a look please or reassign?

**Comment by ahg-g** [comment]
@draveness can you please fix the conflicts in the compatibility test.

**Comment by deads2k** [comment]
It seems weird to have a gate and not use the value.  It's also weird to have a value set one way for testing and another way to actually run.  That was just in the top part of the code.

Which packages are asking for my review?

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 12]
```diff
@@ -294,11 +294,14 @@ type Controller struct {
...
-	taintNodeByCondition bool
-
 	nodeUpdateQueue workqueue.Interface
+
+	// if set to true, Controller will skip taint nodes by condition.
+	//
+	// NOTE: used in scheduler integration test.
+	// TODO(draveness): remove this after the scheduler integration test TestTaintBasedEvictions
+	// could pass with TaintNodesByCondition enabled.
+	SkipTaintNodeByCondition bool
```
Comment (ahg-g) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: why is this needed? why isn't TestTaintBasedEvictions passing without this?

Comment (draveness) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: The tests would fail if we enable TaintNodeByConditions https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82703/pull-kubernetes-integration/1174931892722470913

File: test/integration/scheduler/taint_test.go [hunk 13]
```diff
@@ -689,13 +690,14 @@ func TestTaintBasedEvictions(t *testing.T) {
 				0.55,             // Unhealthy zone threshold
 				true,             // Run taint manager
 				true,             // Use taint based evictions
-				false,            // Enabled TaintNodeByCondition feature
```
Comment (draveness) [inline test/integration/scheduler/taint_test.go]: @ahg-g TestTaintBasedEvictions needs to disable the TaintNodeByCondition here

Comment (ahg-g) [reply test/integration/scheduler/taint_test.go]: We shouldn't be testing code paths that can only be enabled in tests. The test should be re-done to create the desired test state without needing to manually disable parts of the code.

@Huang-Wei I think you did work on this feature (TaintBasedEvictions), can you please take a look and advise how we can fix this? How should the test be re-written to test TaintBasedEvictions while TaintNodeByCondition is enabled.

Comment (Huang-Wei) [reply test/integration/scheduler/taint_test.go]: A little background on this:

I used to write an e2e test for TaintBasedEviction, which works well, but only on GCE/GKE.

So @ravisantoshgudimetla and @damemi proposed to convert that e2e test to an integration one, which can be portable on all providers. The integration is the current version of taint_test.go#TestTaintBasedEvictions. I reviewed and approved that PR #81856, but overlooked it set the TaintNodesByCondition flag to false.

I totally agree the test shouldn't depend on the false flag, I will take a close look to make the integration test work (without setting the flag to false) first.

Comment (ahg-g) [reply test/integration/scheduler/taint_test.go]: can we revert the test in this PR, and @Huang-Wei work in parallel to add it back without the flag so that we don't block this PR? it was not practically useful anyway since it didn't test the actual setup we have right now for the feature flags.

Comment (Huang-Wei) [reply test/integration/scheduler/taint_test.go]: @ahg-g I just proposed a PR #84036 solving that issue. After it's merged, @draveness should be able to proceed on this one.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 14]
```diff
@@ -1198,12 +1198,122 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+				"LeastRequestedPriority",
+				"BalancedResourceAllocation",
+				"SelectorSpreadPriority",
+				"NodePreferAvoidPodsPriority",
+				"NodeAffinityPriority",
+				"TaintTolerationPriority",
+				"InterPodAffinityPriority",
+				"MostRequestedPriority",
+				"RequestedToCapacityRatioPriority",
+			),
```
Comment (ahg-g) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: I think you need to add wantPlugins here.

**Comment by ahg-g** [comment]
> It seems weird to have a gate and not use the value. It's also weird to have a value set one way for testing and another way to actually run. That was just in the top part of the code.
> 
> Which packages are asking for my review?

@deads2k thanks again for catching this, the issue is now addressed. PTAL.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 15]
```diff
@@ -47,12 +47,9 @@ func defaultPredicates() sets.String {
 		predicates.MatchInterPodAffinityPred,
 		predicates.NoDiskConflictPred,
 		predicates.GeneralPred,
-		predicates.CheckNodeMemoryPressurePred,
```
Comment (deads2k) [inline pkg/scheduler/algorithmprovider/defaults/defaults.go]: my approval doesn't cover this.  Commenting so the scheduling guys notice.

File: pkg/scheduler/algorithmprovider/plugins_test.go [hunk 16]
```diff
@@ -81,8 +78,8 @@ func TestApplyFeatureGates(t *testing.T) {
 				t.Fatalf("Error retrieving provider: %v", err)
 			}
 
-			if !p.FitPredicateKeys.Has("CheckNodeCondition") {
-				t.Fatalf("Failed to find predicate: 'CheckNodeCondition'")
+			if p.FitPredicateKeys.Has("CheckNodeCondition") {
```
Comment (deads2k) [inline pkg/scheduler/algorithmprovider/plugins_test.go]: expected?

**Comment by deads2k** [comment]
/approve

controller, admission, and kubeapiserver changes all look fine.

File: pkg/scheduler/algorithmprovider/plugins_test.go (hunk 16 above)
Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/plugins_test.go]: it is fine, we could remove it completely since it is not guarded by feature gate anymore.

File: pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 15 above)
Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/defaults/defaults.go]: yes, those are all replaced by the new feature.

**Comment by draveness** [comment]
/assign @tallclair 
for kubelet approval

**Comment by wojtek-t** [comment]
/approve

/hold
Please rebase and hold cancel.

**Comment by draveness** [comment]
/hold

The integration tests is failing, I'll fix this.

Regards,
Draven
